src/lib/zkapps_examples/tokens/zkapps_tokens.ml

```diff
@@ -0,0 +1,724 @@
+          Field.equal (fst new_stack) (Field.constant empty_hash)
+        in
+        Boolean.(
+          Assert.any
+            [ is_correct; is_empty &&& forest_is_empty &&& new_stack_is_empty ]) ;
+        (elt, new_stack)
+    end
+
+    module State = struct
+      module Value = struct
+        type t =
+          { forest : Zkapp_call_forest.t
+          ; pending_forests : Merkle_list.Value.t
+          ; check_may_use_token : Account_update.May_use_token.t
+          }
+        [@@deriving hlist]
+      end
+
+      module Circuit = struct
+        type t =
+          { forest : Zkapp_call_forest.Checked.t
+          ; pending_forests : Merkle_list.Circuit.t
+          ; check_may_use_token : Account_update.May_use_token.Checked.t
+          }
+        [@@deriving hlist]
+
+        let create forest check_may_use_token =
+          { forest
+          ; pending_forests =
+              ( Field.constant Merkle_list.empty_hash
+              , Prover_value.create (fun () -> []) )
+          ; check_may_use_token
+          }
+      end
+
+      let typ =
+        Typ.of_hlistable
+          [ Zkapp_call_forest.typ
+          ; Merkle_list.typ
+          ; Account_update.May_use_token.typ
+          ]
+          ~var_to_hlist:Circuit.to_hlist ~var_of_hlist:Circuit.of_hlist
+          ~value_to_hlist:Value.to_hlist ~value_of_hlist:Value.of_hlist
+    end
+
+    let dummy_proof =
+      lazy (Pickles.Proof.dummy Nat.N2.n Nat.N2.n Nat.N2.n ~domain_log2:15)
+
+    let next_account_update
+        ({ forest; pending_forests; check_may_use_token } : State.Circuit.t) :
+        State.Circuit.t
+        * Zkapp_call_forest.Checked.account_update
+        * Account_update.May_use_token.Checked.t =
+      let dummy_account_update_body = Lazy.force dummy_account_update_body in
+      let dummy : _ Zkapp_command.Call_forest.Tree.t =
+        { account_update =
+            { Account_update.body = dummy_account_update_body.data
+            ; authorization = Control.None_given
+            }
+        ; account_update_digest = dummy_account_update_body.hash
+        ; calls = []
+        }
+      in
+      let dummy_tree_hash =
+        Zkapp_command.Digest.Tree.constant (Lazy.force dummy_tree_hash)
+      in
+      let (account_update, forest), rest_of_forest =
+        Zkapp_call_forest.Checked.pop ~dummy ~dummy_tree_hash forest
+      in
+      let rest_of_forest_is_empty =
+        Zkapp_call_forest.Checked.is_empty rest_of_forest
+      in
+      let pending_forests =
+        let new_pending_forests =
+          Merkle_list.push
+            { forest = rest_of_forest; may_use_token = check_may_use_token }
+            pending_forests
+        in
+        Merkle_list.if_ rest_of_forest_is_empty ~then_:pending_forests
+          ~else_:new_pending_forests
+      in
+      ( { forest
+        ; pending_forests
+        ; check_may_use_token =
+            Account_update.May_use_token.Checked.constant Inherit_from_parent
```

Comment:
I'm having a hard time seeing whether `check_may_use_token` in the state / Merkle list of pending forests is ever set to anything different than this constant.

Reply:
If it can't be anything else, would really help readability to either comment on this fact or simply remove it from all the data structures

Reply:
For the top level of the tree, it's `Parents_own_token`, but then every lower layer is `Inherit_from_parent`.

Reply:
ah, that makes sense!